src/openssl/mod.rs

```diff
@@ -310,20 +325,358 @@ impl ECVRF {
             }
         });
 
-        if let Some(pt) = point.as_mut() {
-            let mut new_pt = EcPoint::new(&self.group.as_ref())?;
-            new_pt.mul(
-                &self.group.as_ref(),
-                &pt,
-                &BigNum::from_slice(&[self.cofactor])?.as_ref(),
-                &self.bn_ctx,
-            )?;
-            *pt = new_pt;
+        if self.cofactor != 1 {
+            if let Some(pt) = point.as_mut() {
+                let mut new_pt = EcPoint::new(&self.group.as_ref())?;
+                new_pt.mul(
+                    &self.group.as_ref(),
+                    &pt,
+                    &BigNum::from_slice(&[self.cofactor])?.as_ref(),
+                    &self.bn_ctx,
+                )?;
+                *pt = new_pt;
+            }
         }
         // Return error if no valid point was found
         point.ok_or(Error::HashToPointError)
     }
 
+    // Utility function used in hash-to-curve methods.
+    // For security reasons, all field operations, comparisons, and assignments MUST be implemented in constant time
+    // (i.e., execution time MUST NOT depend on the values of the inputs), and without branching.
+
+    /// Function to move a field element conditionally.
+    /// Stated in [hash-to-curve-04](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-4)
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `a` - A field element
+    /// * `b` - A field element
+    /// * `c` - A boolean value
+    ///
+    /// # Returns
+    ///
+    /// * If c is False, CMOV returns a otherwise b
+    fn cmov(a: BigNum, b: BigNum, c: bool) -> BigNum {
+        // It has not been confirmed to have constant time
+        if c {
+            b
+        } else {
+            a
+        }
+    }
+
+    fn icmov(a: i32, b: i32, c: bool) -> i32 {
+        // It has not been confirmed to have constant time
+        if c {
+            b
+        } else {
+            a
+        }
+    }
+
+    /// Function to check whether the given field element is square or not.
+    /// Stated in [hash-to-curve-04](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-4)
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * Returns true whenever the value x is a square in the field given F otherwise false.
+
+    #[allow(clippy::wrong_self_convention)]
+    fn is_square(&mut self, x: &BigNum) -> Result<bool, Error> {
+        let one = BigNum::from_u32(1)?;
+        let two = BigNum::from_u32(2)?;
+
+        let mut inv2 = BigNum::new()?;
+        inv2.mod_inverse(&two, &self.p, &mut self.bn_ctx)?;
+
+        let mut qm1 = BigNum::new()?;
+        qm1.mod_sub(&self.p, &one, &self.p, &mut self.bn_ctx)?;
+        let mut qm1d2 = BigNum::new()?;
+        qm1d2.mod_mul(&qm1, &inv2, &self.p, &mut self.bn_ctx)?;
```

Comment:
Calling `is_square` twice means you need to calculate `qm1d2` twice, which is unneccessary. How do you think about it? This is just a question.

Reply:
We can optimize by making `is_square` to use pre-calculated `qm1d2` value but I'm not worrying about the cost of that calculation. When the performance really matters, I can optimize it later including other similar constants in `sqrt` and `sgn0`.

Reply:
Umm, never mind. Let's go as now.